starboard/shared/libiamf/iamf_config_reader.cc

```diff
@@ -0,0 +1,531 @@
+    }
+  }
+
+  if (error) {
+    return -1;
+  }
+  return i + 1;
+}
+
+int IamfConfigReader::BufferReader::ReadStringInternal(const uint8_t* buf,
+                                                       std::string* str) {
+  SB_DCHECK(buf);
+
+  int remaining_size = static_cast<int>(size_) - pos_;
+  if (remaining_size <= 0) {
+    return -1;
+  }
+
+  // The size of the string is capped to 128 bytes.
+  int max_str_size = std::min(remaining_size, 128);
+  str->clear();
+
+  size_t size = 0;
+  while (buf[size] != '\0' && size < max_str_size) {
+    size++;
+  }
+
+  if (buf[size] != '\0') {
+    return -1;
+  }
+
+  if (size > 0) {
+    str->resize(size);
+    std::memcpy(str->data(), reinterpret_cast<const char*>(buf), size);
+  }
+
+  // Account for null terminator byte.
+  return ++size;
+}
+
+IamfConfigReader::IamfConfigReader(
+    const scoped_refptr<InputBuffer>& input_buffer,
+    bool prefer_binaural_audio,
+    bool prefer_surround_audio)
+    : prefer_binaural_audio_(prefer_binaural_audio),
+      prefer_surround_audio_(prefer_surround_audio) {
+#if SB_IS_BIG_ENDIAN
+#error IamfConfigReader assumes little-endianness.
+#endif  // SB_IS_BIG_ENDIAN
+  SB_DCHECK(!(prefer_binaural_audio && prefer_surround_audio));
+  Read(input_buffer);
+}
+
+bool IamfConfigReader::Read(const scoped_refptr<InputBuffer>& input_buffer) {
+  SB_DCHECK(input_buffer->data());
+  BufferReader reader(input_buffer->data(), input_buffer->size());
+
+  while (!is_valid() && reader.pos() < reader.size()) {
+    RCHECK(ReadOBU(&reader));
+  }
+
+  if (reader.error()) {
+    return false;
+  }
+
+  data_size_ = reader.size() - config_size_;
+  config_obus_.assign(reader.buf(), reader.buf() + config_size_);
+  data_.assign(reader.buf() + config_size_, reader.buf() + reader.size());
+
+  return true;
+}
+
+bool IamfConfigReader::ReadOBU(BufferReader* reader) {
+  SB_DCHECK(reader);
+  uint8_t obu_type = 0;
+  uint32_t obu_size = 0;
+  if (!ReadOBUHeader(reader, &obu_type, &obu_size)) {
+    SB_DLOG(ERROR) << "Error reading OBU header";
+    return false;
+  }
+
+  int next_obu_pos = reader->pos() + obu_size;
+
+  auto skip_param_definition = [&]() {
+    // parameter_id
+    RCHECK(reader->SkipLeb128());
+    // parameter_rate
+    RCHECK(reader->SkipLeb128());
+    uint8_t param_definition_mode;
+    RCHECK(reader->Read1(&param_definition_mode));
+    param_definition_mode = param_definition_mode >> 7;
+    if (param_definition_mode == static_cast<uint8_t>(0)) {
+      // duration
+      RCHECK(reader->SkipLeb128());
+      uint32_t constant_subblock_duration;
+      RCHECK(reader->ReadLeb128(&constant_subblock_duration));
+      if (constant_subblock_duration == 0) {
+        uint32_t num_subblocks;
+        RCHECK(reader->ReadLeb128(&num_subblocks));
+        for (int i = 0; i < num_subblocks; ++i) {
+          // subblock_duration
+          RCHECK(reader->SkipLeb128());
+        }
+      }
+    }
+    return true;
+  };
+
+  switch (static_cast<int>(obu_type)) {
```

Comment:
The switch is too long. Let's add separate functions for parsing obu info of different obu type.